src/Features/CSharp/Portable/DocumentSymbols/CSharpDocumentSymbolsService.cs

```diff
@@ -0,0 +1,248 @@
+                                      SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier);
+
+        private static readonly ImmutableArray<string> s_constantTag = ImmutableArray.Create(WellKnownTags.Constant);
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpDocumentSymbolsService()
+        {
+        }
+
+        protected override ISymbol? GetSymbol(SemanticModel semanticModel, SyntaxNode node, CancellationToken cancellationToken)
+            => node switch
+            {
+                NamespaceDeclarationSyntax or GlobalStatementSyntax => null,
+                MemberDeclarationSyntax member => semanticModel.GetDeclaredSymbol(member, cancellationToken),
+                LabeledStatementSyntax labeledStatement => semanticModel.GetDeclaredSymbol(labeledStatement, cancellationToken),
+                SingleVariableDesignationSyntax singleVariable => semanticModel.GetDeclaredSymbol(singleVariable, cancellationToken),
+                VariableDeclaratorSyntax variableDeclarator => semanticModel.GetDeclaredSymbol(variableDeclarator, cancellationToken),
+                LocalFunctionStatementSyntax localFunction => semanticModel.GetDeclaredSymbol(localFunction, cancellationToken),
+                TypeParameterSyntax typeParameter => semanticModel.GetDeclaredSymbol(typeParameter, cancellationToken),
+                _ => null,
+            };
+
+        protected override bool ShouldSkipSyntaxChildren(SyntaxNode node, DocumentSymbolsOptions options)
+        {
+            // If we're not looking for the full hierarchy, we don't care about things nested inside type members
+            if (options == DocumentSymbolsOptions.TypesAndMembersOnly)
+            {
+                return node is BaseMethodDeclarationSyntax or BasePropertyDeclarationSyntax
+                            or BaseFieldDeclarationSyntax or StatementSyntax or ExpressionSyntax;
+            }
+
+            // We don't return info about things nested inside lambdas for simplicity
+            return node is LambdaExpressionSyntax or CastExpressionSyntax
+                        or AttributeListSyntax or TypeParameterConstraintClauseSyntax;
+        }
+
+        protected override DocumentSymbolInfo GetMemberInfoForType(INamedTypeSymbol type, SyntaxTree tree, ISymbolDeclarationService declarationService, CancellationToken cancellationToken)
+        {
+            using var _ = ArrayBuilder<DocumentSymbolInfo>.GetInstance(out var membersBuilder);
+
+            foreach (var member in type.GetMembers())
+            {
+                switch (member)
+                {
+                    case { IsImplicitlyDeclared: true } or { Kind: SymbolKind.NamedType } or
+                         IMethodSymbol { MethodKind: MethodKind.PropertyGet or MethodKind.PropertySet }:
+                        continue;
```

Comment:
can you break these into separate cases, and comment as to why you're doing this?